overlord/fdestate/fdestate.go

```diff
@@ -634,3 +636,123 @@ func SystemEncryptedFromState(st *state.State) (bool, error) {
 	}
 	return encrypted, nil
 }
+
+type volumesAuthOptionsKey struct{}
+
+// ReplaceProtectedKey creates a taskset that replaces the
+// protected key for the specified target key slots.
+//
+// If keyslotRefs is empty, the following key slots are targets:
+//   - container-role: system-data, name: default
+//   - container-role: system-data, name: default-fallback
+//   - container-role: system-save, name: default-fallback
+func ReplaceProtectedKey(st *state.State, volumesAuth *device.VolumesAuthOptions, keyslotRefs []KeyslotRef) (*state.TaskSet, error) {
+	authMode := device.AuthModeNone
+	if volumesAuth != nil {
+		if err := volumesAuth.Validate(); err != nil {
+			return nil, err
+		}
+		authMode = volumesAuth.Mode
+	}
+
+	if len(keyslotRefs) == 0 {
+		// by default, target protected keys that would have been added during installation.
+		keyslotRefs = append(keyslotRefs,
+			KeyslotRef{ContainerRole: "system-data", Name: "default"},
+			KeyslotRef{ContainerRole: "system-data", Name: "default-fallback"},
+			KeyslotRef{ContainerRole: "system-save", Name: "default-fallback"},
+		)
+	}
+
+	tmpKeyslotRefs := make([]KeyslotRef, 0, len(keyslotRefs))
+	tmpKeyslotRenames := make(map[string]string, len(keyslotRefs))
+	for _, keyslotRef := range keyslotRefs {
+		if err := keyslotRef.Validate(); err != nil {
+			return nil, fmt.Errorf("invalid key slot reference %s: %v", keyslotRef.String(), err)
+		}
+		// TODO:FDEM: accept custom key slot names when a naming convension is defined
+		if keyslotRef.Name != "default" && keyslotRef.Name != "default-fallback" {
+			return nil, fmt.Errorf(`invalid key slot reference %s: unsupported name, expected "default" or "default-fallback"`, keyslotRef.String())
+		}
+
+		tmpKeyslotRef := tmpKeyslotRef(keyslotRef)
+		tmpKeyslotRefs = append(tmpKeyslotRefs, tmpKeyslotRef)
+		tmpKeyslotRenames[tmpKeyslotRef.String()] = keyslotRef.Name
+	}
+
+	unlockedWithRecoveryKey, err := boot.IsUnlockedWithRecoveryKey()
+	if err != nil && !errors.Is(err, fs.ErrNotExist) {
+		return nil, err
+	}
+	if unlockedWithRecoveryKey {
+		// primary key will be missing from kernel keyring if disk was
+		// unlocked with recovery key during boot.
```

Comment:
It might not always be missing. If we use a protector key token for the save disk, it will find the primary key.